apollo-federation/src/query_plan/operation.rs

```diff
@@ -2527,8 +2554,134 @@ impl SelectionSet {
         self.selections.contains_key(key)
     }
 
-    pub(crate) fn add_at_path(&mut self, path: &OpPath, selection_set: Option<&Arc<SelectionSet>>) {
-        Arc::make_mut(&mut self.selections).add_at_path(path, selection_set)
+    /// Inserts a `Selection` into the inner map. Should a selection with the same key already
+    /// exist in the map, the existing selection and the given selection are merged, replacing the
+    /// existing selection while keeping the same insertion index.
+    fn add_selection(
+        &mut self,
+        parent_type: &CompositeTypeDefinitionPosition,
+        schema: &ValidFederationSchema,
+        selection: Selection,
+    ) -> Result<(), FederationError> {
+        let selections = Arc::make_mut(&mut self.selections);
+
+        let key = selection.key();
+        match selections.remove(&key) {
+            Some((index, existing_selection)) => {
+                let to_merge = [existing_selection, selection];
+                // `existing_selection` and `selection` both have the same selection key,
+                // so the merged selection will also have the same selection key.
+                let selection = SelectionSet::make_selection(schema, parent_type, to_merge.iter())?
+                    .ok_or_else(|| {
+                        FederationError::internal("Failed to merge selections together.")
+                    })?;
+                selections.insert_at(index, selection);
+            }
+            None => {
+                selections.insert(selection);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Adds a path, and optional some selections following that path, to this selection map.
+    ///
+    /// Today, it is possible here to add conflicting paths, such as:
+    /// - `add_at_path("field1(arg: 1)")`
+    /// - `add_at_path("field1(arg: 2)")`
+    ///
+    /// Users of this method should guarantee that this doesn't happen. Otherwise, converting this
+    /// SelectionSet back to an ExecutableDocument will return a validation error.
+    ///
+    /// The final selections are optional (for instance, if `path` ends on a leaf field,
+    /// then no followup selections would make sense),
+    /// but when some are provided, uncesssary fragments will be automaticaly removed
```

Comment:
typo: "uncesssary" and "automaticaly". I know it's just copied from JS code. It may be a good time to correct it :)
I use Code Spell Checker VS Code extension. I think it's pretty good.